dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CachingFileTree.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2024 OICR and UCSC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.dockstore.webservice.helpers;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Implements a FileTree that wraps a specified FileTree, delegates to
+ * the underlying FileTree's methods, saves the returned values, and
+ * returns the saved value when a method is invoked multiple times with
+ * the same arguments.
+ *
+ * For example, the first call to `readFile("/foo.txt")` is delegated
+ * to the underlying FileTree, and the resulting content is saved and
+ * returned.  For subsequent calls to `readFile("/foo.txt")`, the saved
+ * content is returned, and the underlying FileTree is not used.
+ *
+ * Useful for avoiding repeated accesses to a FileTree-abstracted
+ * resource when the code makes multiple passes over the same files,
+ * without needing to explicitly propagate the retrieved information
+ * (by passing it down the call stack, saving it in a variable, etc).
+ */
+public class CachingFileTree implements FileTree {
+
+    private final FileTree fileTree;
+    /**
+     * Maps file paths to saved file content.
+     */
+    private final Map<String, String> filePathToContent = new HashMap<>();
```

Comment:
This functionally looks like a Guava Loading cache 
https://github.com/google/guava/wiki/cachesexplained#example

There may be other cache implementations floating around our classpath. 
Ditto with dirPathToFiles

Reply:
While at it, it's a bit confusing because the map is from string to string.
Consider using a Path https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/nio/file/Path.html as the key so things are more typed.

Reply:
Misc observations:

* Yes, it's a very simple cache based on a `Map`, also on our classpath.  The main advantage to using a full-blown loading cache probably would be to reduce the attack surface slightly (harder to DOS), but at the expense of complexity, performance, and a larger chance that we'll inadvertently use it incorrectly.  If someone feels really strongly, we can do it, though...

* `Path` has system-dependent behavior.  For example, the file separator changes from system to system.  Because of that, would actually recommend, in general, that `Path` be avoided when possible, except for purposes such as accessing the local file system. Aspirational, of course...

* While the semantics of `Path` and the semantics of workflow file paths seem to overlap each other 1:1 (more or less), it's not implausible that there's some weird file syntax or structure that we'd want to support someday that `Path` doesn't. 

* In our codebase, you will find `Map<String, String>` all over the place, and we very often use `String` to refer to paths.  So, it meets the standard of our existing code, at least.  Regarding this particular situation, don't think the advantage of `Path` in the type outweighs the ick of converting the incoming value, which is a `String`, to a `Path` in each method.

Reply:
> So, it meets the standard of our existing code, at least.  

FWIW, while it is true existing code is like this, this is one of the regrets I've had with the existing code, so if we're improving things or writing a new path, I'd like the make things more typesafe/more easily readable.

> it's not implausible that there's some weird file syntax or structure that we'd want to support someday that Path doesn't.

Seems unlikely, but in that case it seems like you can implement `Path` anew, " WARNING: This interface is only intended to be implemented by those developing custom file system implementations. Methods may be added to this interface in future releases. " 
https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html

> Path has system-dependent behavior.

I think I'm ok with Dockstore always running on Linux or Linux-like systems?

> The main advantage to using a full-blown loading cache probably would be to reduce the attack surface slightly (harder to DOS)

On the contrary, I think the main advantage is that we'd avoid rolling our own cache implementation. We'd gain access to other pre-made cache functionality in Guava (or similar) if we need to tweak things, and it's easier to understand because instead of it being our own cache implementation it's a well-known one.